store/postgres/src/chain_store.rs

```diff
@@ -952,38 +954,67 @@ mod data {
             conn: &mut PgConnection,
             block_ptr: BlockPtr,
             offset: BlockNumber,
+            root: Option<BlockHash>,
         ) -> Result<Option<(json::Value, BlockPtr)>, Error> {
-            let data_and_hash = match self {
+            let short_circuit_predicate = if root.is_some() {
+                "and b.parent_hash <> $3"
+            } else {
+                ""
+            };
+
+            let data_and_ptr = match self {
                 Storage::Shared => {
-                    const ANCESTOR_SQL: &str = "
+                    let query = format!("
         with recursive ancestors(block_hash, block_offset) as (
             values ($1, 0)
             union all
             select b.parent_hash, a.block_offset+1
               from ancestors a, ethereum_blocks b
              where a.block_hash = b.hash
                and a.block_offset < $2
+               {}
         )
-        select a.block_hash as hash
+        select a.block_hash as hash, b.number as number
           from ancestors a
-         where a.block_offset = $2;";
+               inner join ethereum_blocks b on a.block_hash = b.hash
```

Comment:
This join is probably not a big deal, but couldn't it be avoided if the recursive CTE also selected the block number, i.e. became `with recursive ancestors (block_hash, block_number, block_offset) ..` ?